lib/eslint/eslint-helpers.js

```diff
@@ -151,94 +285,185 @@ async function globSearch({ basePath, patterns, configs }) {
+             * first matching pattern for improved speed.
+             */
+            const matchesPattern = unmatchedPatterns.size > 0
+                ? matchers.reduce((previousValue, matcher) => {
+                    const pathMatches = matcher.match(relativePath);
+
+                    /*
+                     * We updated the unmatched patterns set only if the path
+                     * matches and the file isn't ignored. If the file is
+                     * ignored, that means there wasn't a match for the
+                     * pattern so it should not be removed.
+                     *
+                     * Performance note: isFileIgnored() aggressively caches
+                     * results so there is no performance penalty for calling
+                     * it twice with the same argument.
+                     */
+                    if (pathMatches && !configs.isFileIgnored(entry.path)) {
+                        unmatchedPatterns.delete(matcher.pattern);
+                    }
+
+                    return pathMatches || previousValue;
+                }, false)
+                : matchers.some(matcher => matcher.match(relativePath));
 
             return matchesPattern && !configs.isFileIgnored(entry.path);
         }
+
     })).map(entry => entry.path);
 
+    // now check to see if we have any unmatched patterns
+    if (errorOnUnmatchedPattern && unmatchedPatterns.size > 0) {
+        throw new UnmatchedSearchPatternsError({
+            basePath,
+            unmatchedPatterns: [...unmatchedPatterns].map(
+                pattern => relativeToPatterns.get(pattern)
+            ),
+            patterns,
+            rawPatterns
+        });
+    }
+
+    return filePaths;
+}
+
+/**
+ * Checks to see if there are any ignored results for a given search. This
+ * happens either when there are unmatched patterns during a search or if
+ * a search returns no results.
+ * @param {Object} options The options for this function.
+ * @param {string} options.basePath The directory to search.
+ * @param {Array<string>} options.patterns An array of glob patterns
+ *      that were used in the original search.
+ * @param {Array<string>} options.rawPatterns An array of glob patterns
+ *      as the user inputted them. Used for errors.
+ * @param {Array<string>} options.patternsToCheck An array of glob patterns
+ *      to use for this check.
+ * @param {boolean} options.errorOnUnmatchedPattern Determines if an error
+ *      should be thrown when a pattern is unmatched.
+ * @returns {void}
+ * @throws {NoFilesFoundError} If there is a pattern that doesn't match
+ *      any files and `errorOnUnmatchedPattern` is true.
+ * @throws {AllFilesIgnoredError} If there is a pattern that matches files
+ *      when there are no ignores.
+ */
+async function checkForIgnoredResults({
+    basePath,
+    patterns,
+    rawPatterns,
+    patternsToCheck = patterns,
+    errorOnUnmatchedPattern
```

Comment:
Related to https://github.com/eslint/eslint/pull/16462#discussion_r1009869403, it doesn't seem that this function should  have `errorOnUnmatchedPattern` param, but rather be called only if `errorOnUnmatchedPattern = true`?